CDCgov/prime-reportstream
# Victor/7427 hl7 errors

This PR adds the ability for UP return specific errors when bad HL7 data is submitted. Currently our implementation of the HL7Reader has validation turned off and will only throw an error if it fails to parse the HL7 message. Because of this the amount of bad data you can send is very minimal. Once validation is turned on we can check if more specific error messages need to be added.

Test Steps:
1. Submit bad data to the reports API. Sample bad data can be found in the unit tests added to `HL7ReaderTests.kt` on this PR.
2. Test required field is missing. Sample  HL7 message can be found in unit test `'test decoding of bad HL7 message - MSH Version ID missing'`
3. API should return this error message: "Required field missing: Can't find version ID - MSH.12 is null"
4. Test  Data type error. Sample  HL7 message can be found in unit test `test decoding of bad HL7 message - OBX Wrong data type`
5. API should return this error message: "Data type error: 'test' in record 1 is invalid for version 2.5.1 at OBX-2(0)"
6. I tried sending other types of bad data, but nothing would throw an error. I tried sending duplicate lines, taking out required fields, bad data for dates, gender, and country codes.

## Changes
- Added logic to `HL7Reader.kt` to log and return more specific errors when bad HL7 data is submitted.

## Checklist

### Testing

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #7427
